심화/Chapter 1/소민/README.md

```diff
@@ -0,0 +1,172 @@
+# DB 부하 줄이기
+
+## 데이터 자체를 쪼개기
+
+### Sharding
+<details><summary>샤딩 구조</summary>
+
+[![샤딩 이미지](./images/sharding.png)](./images/sharding.png)
+
+</details>
+
+하나의 DB를 개별 서버의 샤드로 나누는 것을 의미한다.
+* 샤드란? 데이터를 나눈 작은 단위
+* 즉 데이터를 나눈 작은 단위(샤드)를 개별 DB 서버에서 사용하는 것
+
+같은 스키마(테이블 구조)를 가진 샤드를 여러대로 분리한다는 점이 특징
+* 동일한 테이블을 여러 DB 서버(샤드)로 나누어 트래픽을 분산시키는 수평적 확장
+* 개별 샤드에 동일한 스키마의 중복되지 않는 데이터가 저장되어 있다.
+
+#### 장점
+* 여러 서버에 데이터가 분산 저장되어 쿼리 성능 향상 (필요한 데이터가 있는 샤드에만 접근하므로, 스캔 범위도 줄어들고 병렬적으로 여러 샤드 스캔이 가능함)
+* 개별 샤드에 대해 병렬적으로 RW 가능 -> 쓰기 성능 향상
+* 단일 DB의 용량 문제도 해결할 수 있음 (대용량 데이터 분리) 
+
+#### 단점
+* 전체 데이터에 대한 쿼리 복잡도 향상
+  * e.g., 전체 데이터 개수를 조회하는 쿼리
+* 샤드 간 트랜잭션 관리 복잡도 향상
+
+#### 장애 복구
+* 문제가 생긴 샤드만 복구
+* replication/백업된게 있으면 바로 복구 가능, 그 외에는 다운타임 발생(?)
+
+
+### Partitioning
+
+<details><summary>파티셔닝 구조</summary>
+
+[![파티셔닝 이미지](./images/partitioning.png)](./images/partitioning.png)
+
+</details>
+
+샤딩과 비슷하지만 **하나의 DB 서버 내에서 한 테이블의 데이터를 분리하는 것**을 의미한다.
+* 한 테이블에 있는 다량의 데이터를 여러 파티션으로 나눠서 저장하는 것
+* 개별 파티션에 동일한 스키마, 중복되지 않는 데이터가 존재한다.
+
+#### 장점
+* 필요한 데이터가 있는 파티션만 스캔하여 쿼리 -> 성능 향상
+* 여러 파티션에 대해 병렬적 RW 작업 가능 -> 쓰기 성능 향상
+* 단일 DB 서버 한 대로 데이터 분산 가능
+
+#### 단점
+* 샤딩과 마찬가지로, 파티션을 나눌 키/전략 설정이 필요함
+* 하나의 파티션에 데이터가 몰리는 불균형 발생 가능
+* 여러 파티션에 걸쳐있는 데이터를 조회해야 하는 쿼리 복잡도 향상
+
+#### 장애 복구
+* 백업이 있으면 해당 백업본으로 복구 -> 데이터 양이 많다면 복구가 오래걸릴 수 있음
+* MySQL 등 일부 DBMS에서는 파티션에 대한 replication도 제공해서 바로 복구 가능
+
+
+## 데이터를 쪼개지 않고 성능 향상하기
+
+
+### Replication
+
+<details><summary>레플리케이션 구조</summary>
+
+[![레플리케이션 이미지](./images/replication.png)](./images/replication.png)
+
+</details>
+
+DB 다중화로 **같은 스키마, 같은 데이터의 DB 서버를 여러 대** 두어 사용하는 것.
+* 동일 데이터를 여러 DB 서버에 복제
+
+master-slave 구조로 활용한다.
+* 쓰기 작업은 master DB에 대해 수행되고, slave DB는 읽기 전용으로 사용된다.
+  * Read 성능은 향상 가능하지만, Write 성능은 크게 향상되지 않는다.
```

Comment:
호옹 이유가 궁금해요 소민 선배님 >-<

Reply:
MySQL 공식 문서에는 master-slave 구조의 replication 구조를 사용했을 때의 이점이 다음과 같이 나와있다.
- write 성능 향상: master 서버는 오로지 update만 담당하므로
- read의 드라마틱한 성능 향상: read는 slave 서버 수 만큼 병렬 처리가 가능하므로

> Scale-out solutions - spreading the load among multiple replicas to improve performance. In this environment, all writes and updates must take place on the source server. Reads, however, may take place on one or more replicas. This model can improve the performance of writes (since the source is dedicated to updates), while dramatically increasing read speed across an increasing number of replicas.

- [MySQL Replication](https://dev.mysql.com/doc/refman/8.4/en/replication.html)

따라서, write 성능도 함께 향상이 되는 것은 맞지만 read에 비해서는 성능 향상이 눈에 띄게 크지 않다 정도로 해석할 수 있을 것 같다.

Reply:
속이 시원~~~~~~~~~~~~~